File: vehicle_sim.m
function [next_state, reward, done] = step(obj, action)
    % Step function for RIS simulation with dynamic car movement
    % Inputs:
    %   obj: Simulation object with properties
    %   action: RIS phase configuration from the agent
    % Outputs:
    %   next_state: Next state observation
    %   reward: Reward for the action
    %   done: Boolean flag indicating if episode is terminated
    
    % Check if episode has already terminated
    if isEpisodeDone(obj) || obj.stepCount >= obj.maxSteps
        done = true;
        next_state = obj.getState();
        reward = 0;
        return;
    end
    
    % Apply RIS phase configuration from action
    ris_phases = action(1:obj.Nr);
    obj.phi = diag(exp(1i * 2 * pi * ris_phases));
    
    % Compute channel matrices based on current positions
    [obj.H_br, obj.H_bt, obj.H_rt] = computeChannelMatrices(obj);
    
    % Compute performance metrics
    [W, obj.FIM] = computeWx(obj);
    covar_matrix = W * W';
    [peb] = obj.calculatePerformanceMetrics(covar_matrix);
    
    % Calculate minimum data rate
    R_min = computeR_min(obj);
    
    % Update the vehicle's position
    if ~isempty(obj.destination) && ~isequal(obj.car_loc, obj.destination)
        % Calculate direction vector
        direction_vector = obj.destination - obj.car_loc;
        distance = norm(direction_vector);
        
        % Check if destination reached
        if distance < (obj.speed * obj.dt)
            % Generate new random destination within boundaries
            obj.destination = [
                randi([obj.boundary(1,1), obj.boundary(2,1)]), 
                randi([obj.boundary(1,2), obj.boundary(2,2)]), 
                0  % Keep Z at ground level
            ];
            
            % Debug output (can be disabled in production)
            if obj.verbose
                fprintf('Car reached destination. New destination: [%.1f, %.1f, %.1f]\n', ...
                    obj.destination(1), obj.destination(2), obj.destination(3));
            end
        else
            % Normalize direction vector and update position
            direction = direction_vector / distance;
            obj.car_loc = obj.car_loc + direction * obj.speed * obj.dt;
            
            % Debug output (can be disabled in production)
            if obj.verbose && mod(obj.stepCount, 100) == 0
                fprintf('Step %d: Car at [%.1f, %.1f, %.1f], PEB: %.6f\n', ...
                    obj.stepCount, obj.car_loc(1), obj.car_loc(2), obj.car_loc(3), peb);
            end
        end
    else
        % Initialize destination if not set
        if isempty(obj.destination)
            obj.destination = [
                randi([obj.boundary(1,1), obj.boundary(2,1)]), 
                randi([obj.boundary(1,2), obj.boundary(2,2)]), 
                0  % Keep Z at ground level
            ];
        end
    end
    
    % Update the target location (if the car is the dynamic target)
    obj.target_loc = obj.car_loc;
    
    % Update time and step counter
    obj.time = obj.time + obj.dt;
    obj.stepCount = obj.stepCount + 1;
    
    % Store trajectory data for visualization
    obj.trajectory(obj.stepCount,:) = obj.car_loc;
    obj.peb_history(obj.stepCount) = peb;
    obj.time_history(obj.stepCount) = obj.time;
    
    % Recompute geometric parameters based on new positions
    [obj.L1, obj.L2, obj.L3, obj.L_proj1, obj.L_proj2, obj.L_proj3, obj.delays, obj.angles] = ...
        computeGeometricParameters(obj);
    
    % Get the new state
    next_state = getState(obj);
    
    % Calculate reward
    reward = obj.computeReward(peb, obj.rate, R_min);
    
    % Check if the episode should be terminated
    done = isEpisodeDone(obj) || (obj.stepCount >= obj.maxSteps);
end

% Helper function to compute channel matrices
function [H_br, H_bt, H_rt] = computeChannelMatrices(obj)
    % Extract locations
    bs_loc = obj.bs_loc;
    ris_loc = obj.ris_loc;
    target_loc = obj.target_loc;
    
    % Calculate geometric parameters if not already calculated
    if isempty(obj.angles)
        [~, ~, ~, ~, ~, ~, ~, angles] = computeGeometricParameters(obj);
        obj.angles = angles;
    end
    
    % Parameters
    fc = obj.fc;           % Carrier frequency
    lambda = 3e8/fc;       % Wavelength
    Nr = obj.Nr;           % Number of RIS elements
    Nt = obj.Nt;           % Number of BS antennas
    
    % Calculate channel matrices using your existing channel model
    % This is a placeholder - replace with your actual implementation
    
    % BS to RIS channel (H_br)
    H_br = channelModel(obj, obj.angles.theta_br, obj.angles.phi_br, Nr, Nt, lambda, bs_loc, ris_loc);
    
    % BS to Target channel (H_bt)
    H_bt = channelModel(obj, obj.angles.theta_bt, obj.angles.phi_bt, 1, Nt, lambda, bs_loc, target_loc);
    
    % RIS to Target channel (H_rt)
    H_rt = channelModel(obj, obj.angles.theta_rt, obj.angles.phi_rt, 1, Nr, lambda, ris_loc, target_loc);
end

% Simplified channel model function (replace with your actual implementation)
function H = channelModel(obj, theta, phi, Nr, Nt, lambda, tx_loc, rx_loc)
    % This function should be replaced with your actual channel model
    
    % Calculate distance
    d = norm(tx_loc - rx_loc);
    
    % Path loss
    path_loss = lambda/(4*pi*d);
    
    % Array steering vectors
    a_r = arraySteering(Nr, theta, phi, lambda);
    a_t = arraySteering(Nt, theta, phi, lambda);
    
    % LOS component
    H_los = a_r * a_t';
    
    % Add fading component (if used in your model)
    if obj.use_fading
        H_fading = (randn(Nr, Nt) + 1i*randn(Nr, Nt))/sqrt(2) * obj.fading_factor;
        H = path_loss * (H_los + H_fading);
    else
        H = path_loss * H_los;
    end
end

% Function to determine if episode is done
function done = isEpisodeDone(obj)
    % Check if car is out of boundaries
    if any(obj.car_loc < obj.boundary(1,:)) || any(obj.car_loc > obj.boundary(2,:))
        done = true;
        return;
    end
    
    % Check if maximum steps reached
    if obj.stepCount >= obj.maxSteps
        done = true;
        return;
    end
    
    % Episode is not done
    done = false;
end

% Function to get current state
function state = getState(obj)
    % Create state vector based on:
    % 1. Car's position
    % 2. Geometric parameters (angles, distances)
    % 3. Current RIS configuration
    
    % Extract diagonal of phi matrix
    phi_diag = diag(obj.phi);
    phi_angles = angle(phi_diag) / (2*pi);
    
    % Combine all state components
    state = [
        obj.car_loc(:)', ...                        % Car position [x, y, z]
        obj.L1, obj.L2, obj.L3, ...                 % Path lengths
        obj.angles.theta_bt, obj.angles.phi_bt, ... % BS-Target angles
        obj.angles.theta_br, obj.angles.phi_br, ... % BS-RIS angles
        obj.angles.theta_rt, obj.angles.phi_rt, ... % RIS-Target angles
        phi_angles(:)'                              % Current RIS configuration
    ];
end

% Function to compute geometric parameters
function [L1, L2, L3, L_proj1, L_proj2, L_proj3, delays, angles] = computeGeometricParameters(obj)
    % Extract locations
    bs_loc = obj.bs_loc;
    ris_loc = obj.ris_loc;
    target_loc = obj.target_loc;
    
    % Direct path: BS to Target
    L1 = norm(bs_loc - target_loc);
    
    % First reflection path: BS to RIS
    L2 = norm(bs_loc - ris_loc);
    
    % Second reflection path: RIS to Target
    L3 = norm(ris_loc - target_loc);
    
    % Projected distances (horizontal plane)
    L_proj1 = norm(bs_loc(1:2) - target_loc(1:2));
    L_proj2 = norm(bs_loc(1:2) - ris_loc(1:2));
    L_proj3 = norm(ris_loc(1:2) - target_loc(1:2));
    
    % Calculate delays
    c = 3e8;  % Speed of light (m/s)
    delays = [L1/c, (L2+L3)/c];
    
    % Calculate angles
    % Elevation angles (with respect to horizontal plane)
    theta_bt = atan2(target_loc(3) - bs_loc(3), L_proj1);
    theta_br = atan2(ris_loc(3) - bs_loc(3), L_proj2);
    theta_rt = atan2(target_loc(3) - ris_loc(3), L_proj3);
    
    % Azimuth angles
    phi_bt = atan2(target_loc(2) - bs_loc(2), target_loc(1) - bs_loc(1));
    phi_br = atan2(ris_loc(2) - bs_loc(2), ris_loc(1) - bs_loc(1));
    phi_rt = atan2(target_loc(2) - ris_loc(2), target_loc(1) - ris_loc(1));
    
    % Package angles
    angles = struct('theta_bt', theta_bt, 'theta_br', theta_br, 'theta_rt', theta_rt, ...
                   'phi_bt', phi_bt, 'phi_br', phi_br, 'phi_rt', phi_rt);
end

% Function to create array steering vector
function a = arraySteering(N, theta, phi, lambda)
    % Element spacing
    d = lambda/2;
    
    % Initialize steering vector
    a = zeros(N, 1);
    
    % Calculate steering vector elements
    for n = 0:N-1
        a(n+1) = exp(1i * 2 * pi * d * n * sin(theta) * cos(phi) / lambda);
    end
end

% Function to compute minimum data rate
function R_min = computeR_min(obj)
    % This is a placeholder - replace with your actual implementation
    % Calculate the minimum achievable data rate based on channel conditions
    
    % Extract channels
    H_bt = obj.H_bt;
    H_br = obj.H_br;
    H_rt = obj.H_rt;
    phi = obj.phi;
    
    % Calculate effective channel
    H_eff = H_bt + H_br * phi * H_rt;
    
    % Calculate SNR (simplified)
    snr_linear = 10^(obj.snr_db/10);
    
    % Calculate rate (bits/s/Hz)
    R_min = log2(1 + snr_linear * norm(H_eff, 'fro')^2);
end

% Function to compute beamformer matrix Wx
function [W, FIM] = computeWx(obj)
    % This is a placeholder - replace with your actual implementation
    % Compute the beamformer matrix and Fisher Information Matrix
    
    % Extract channels
    H_bt = obj.H_bt;
    H_br = obj.H_br;
    H_rt = obj.H_rt;
    phi = obj.phi;
    
    % Calculate effective channel
    H_eff = H_bt + H_br * phi * H_rt;
    
    % Calculate W (simplified)
    W = H_eff' / (norm(H_eff, 'fro') + eps);
    
    % Calculate FIM (simplified)
    snr_linear = 10^(obj.snr_db/10);
    FIM = snr_linear * (W * W');
end

File: vehicle_sim.py
import numpy as np
from dataclasses import dataclass
from typing import Tuple, NamedTuple, Optional

class Angles(NamedTuple):
    azimuth: float
    elevation_angle: float
    elevation_azimuth: float
    aoa_in: float
    aoa_out: float

class ChannelAngles(NamedTuple):
    bs_to_target: Angles
    bs_to_ris: Angles
    ris_to_target: Angles

class Delays(NamedTuple):
    line_of_sight: float
    non_line_of_sight: float

class WirelessChannelSimulation:
    def __init__(self):
        # System Parameters
        self.Nb = 4  # Number of BS antennas (ULA)
        self.Nt = 4  # Number of Vehicle antennas
        self.Nx = 8  # Number of RIS elements in x-direction
        self.Ny = 8  # Number of RIS elements in y-direction
        self.Nr = self.Nx * self.Ny  # Total number of RIS elements
        self.Mb = 64  # Number of subcarriers
        self.N = 2048  # FFT size
        self.B = 20e6  # Bandwidth (20 MHz)
        self.fc = 28e9  # Carrier frequency
        self.c = 3e8  # Speed of light
        self.lambda_ = self.c / self.fc
        self.d = 0.5 * self.lambda_  # Antenna spacing
        
        # Environment Parameters
        self.x_size = 1000  # Environment size in meters
        self.y_size = 1000
        self.z_size = 100
        
        # Channel Parameters
        self.K_dB = 4  # Rician K-factor in dB
        self.K = 10**(self.K_dB/10)
        self.sigma = np.sqrt(1/(2*(self.K+1)))
        self.mu = np.sqrt(self.K/(self.K+1))
        
        # Initialize positions
        self.vehicle_pos = np.array([500, 500, 0])
        self.goal_pos = np.array([1000, 500, 0])
        self.bs_loc = np.array([900, 100, 20])
        self.ris_loc = np.array([200, 300, 40])
        
        # Initialize array positions
        self.bs_array_positions = self._init_bs_array()
        self.vehicle_array_positions = np.zeros((self.Nt, 3))
        self.ris_array_positions = self._init_ris_array()

    def _init_bs_array(self) -> np.ndarray:
        """Initialize BS array positions."""
        positions = np.zeros((self.Nb, 3))
        for i in range(self.Nb):
            positions[i] = self.bs_loc + np.array([0, i * self.d, 0])
        return positions

    def _init_ris_array(self) -> np.ndarray:
        """Initialize RIS array positions."""
        positions = np.zeros((self.Nr, 3))
        idx = 0
        for ix in range(self.Nx):
            for iy in range(self.Ny):
                positions[idx] = self.ris_loc + np.array([ix * self.d, iy * self.d, 0])
                idx += 1
        return positions

    def generate_steering_vector(self, Nant: int, psi: float) -> np.ndarray:
        """Generate steering vector for ULA."""
        k = 2 * np.pi / self.lambda_
        n = np.arange(Nant)
        phase_terms = np.exp(1j * k * self.d * n * np.sin(psi))
        return phase_terms / np.sqrt(Nant)

    def generate_2d_steering_vector(self, Nx: int, phi_a: float, phi_e: float) -> np.ndarray:
        """Generate 2D steering vector for UPA."""
        N2 = Nx * Nx
        a_phi = np.zeros(N2, dtype=complex)
        k = 2 * np.pi / self.lambda_
        
        idx = 0
        for m in range(Nx):
            for n in range(Nx):
                phase_term = np.exp(1j * k * self.d * 
                                  (m * np.sin(phi_a) * np.sin(phi_e) + 
                                   n * np.cos(phi_e)))
                a_phi[idx] = phase_term
                idx += 1
        
        return a_phi / np.sqrt(N2)

    def generate_channels(self, angles: ChannelAngles) -> Tuple[np.ndarray, np.ndarray, np.ndarray]:
        """Generate all channels (BS-Target, BS-RIS, RIS-Target)."""
        # Generate BS-Target channel
        H_bt = self._generate_H_bt(angles.bs_to_target)
        
        # Generate BS-RIS and RIS-Target channels
        H_br = self._generate_H_br(angles.bs_to_ris)
        H_rt = self._generate_H_rt(angles.ris_to_target)
        
        return H_bt, H_br, H_rt

    def _generate_H_bt(self, angles: Angles) -> np.ndarray:
        """Generate BS-Target channel."""
        a_psi_bt = self.generate_steering_vector(self.Nr, angles.aoa_in)
        a_psi_tb = self.generate_steering_vector(self.Nt, angles.aoa_out)
        return np.outer(a_psi_tb, a_psi_bt)

    def _generate_H_br(self, angles: Angles) -> np.ndarray:
        """Generate BS-RIS channel."""
        a_psi_br = self.generate_steering_vector(self.Nb, angles.azimuth)
        a_phi_abr = self.generate_2d_steering_vector(
            int(np.sqrt(self.Nr)), angles.elevation_azimuth, angles.elevation_angle)
        return np.outer(a_phi_abr, a_psi_br)

    def _generate_H_rt(self, angles: Angles) -> np.ndarray:
        """Generate RIS-Target channel."""
        a_psi_rt = self.generate_steering_vector(self.Nt, angles.azimuth)
        a_phi_art = self.generate_2d_steering_vector(
            int(np.sqrt(self.Nr)), angles.elevation_azimuth, angles.elevation_angle)
        return np.outer(a_psi_rt, a_phi_art)

    def update_vehicle_position(self, new_position: np.ndarray) -> None:
        """Update vehicle position and recalculate related parameters."""
        self.vehicle_pos = new_position
        # Update vehicle array positions based on new position
        for i in range(self.Nt):
            self.vehicle_array_positions[i] = self.vehicle_pos + np.array([0, i * self.d, 0])
            
    def simulate_downlink_transmission(self, delays: Delays, W: Optional[np.ndarray] = None) -> Tuple[np.ndarray, np.ndarray, float]:
        """
        Simulate downlink transmission with both LoS and NLoS paths.
        
        Args:
            delays: Delays for LoS and NLoS paths
            W: Optional beamforming matrix
            
        Returns:
            Tuple of H_NLoS, H_LoS channels and gamma_c parameter
        """
        # Generate transmit data for each subcarrier
        x = (np.random.randn(self.Mb) + 1j * np.random.randn(self.Mb)) / np.sqrt(2)
        
        # Generate beamforming matrix if not provided
        if W is None:
            W = np.zeros((self.Nb, self.Mb), dtype=complex)
            for i in range(self.Mb):
                w = (np.random.randn(self.Nb) + 1j * np.random.randn(self.Nb)) / np.sqrt(2 * self.Nb)
                W[:, i] = w / np.linalg.norm(w)
        
        # Generate basic channels
        angles = self.compute_geometric_parameters()  # You need to implement this
        H_bt, H_br, H_rt = self.generate_channels(angles)
        
        # Generate LoS and NLoS channels
        H_Los = self._generate_H_Los(H_bt, delays.line_of_sight)
        H_NLos = self._generate_H_NLoS(H_rt, H_br, delays.non_line_of_sight)
        
        return H_NLos, H_Los, 90.0  # gamma_c is set to 90 as in the original code

    def _generate_H_Los(self, H_bt: np.ndarray, tau_l: float) -> np.ndarray:
        """Generate Line of Sight channel matrix."""
        # Generate small-scale fading
        hl = (self.sigma * complex(np.random.randn(), np.random.randn())) + self.mu
        
        # Path loss parameters
        rho_l = 3
        gamma_l = np.sqrt(self.Nb * self.Nt) / np.sqrt(rho_l)
        
        # Generate H_Los for all subcarriers
        H_Los = np.zeros((self.Nt, self.Nr, self.N), dtype=complex)
        
        for n in range(self.N):
            phase = np.exp(1j * 2 * np.pi * self.B * n * tau_l / self.N)
            H_Los[:, :, n] = gamma_l * hl * H_bt * phase
            
        return H_Los

    def _generate_H_NLoS(self, H_rt: np.ndarray, H_br: np.ndarray, tau_nl: float) -> np.ndarray:
        """Generate Non-Line of Sight channel matrix."""
        # Generate small-scale fading
        hnl = (self.sigma * complex(np.random.randn(), np.random.randn())) + self.mu
        
        # Path loss parameters
        rho_nl = 4
        gamma_nl = np.sqrt(self.Nb * self.Nr) / np.sqrt(rho_nl)
        
        # Generate RIS reflection parameters
        rho_r = 1
        theta = 2 * np.pi * np.random.rand(self.Nr)
        u = rho_r * np.exp(1j * theta)
        Phi = np.diag(u)
        
        # Generate H_NLoS for all subcarriers
        H_NLoS = np.zeros((self.Nt, self.Nb, self.N), dtype=complex)
        
        for n in range(self.N):
            phase = np.exp(1j * 2 * np.pi * self.B * n * tau_nl / self.N)
            H_NLoS[:, :, n] = gamma_nl * hnl * H_rt @ Phi @ H_br * phase
            
        return H_NLoS
    
    def compute_geometric_parameters():
        # Define locations
        target_loc = np.array([500, 500, 0])  # Vehicle starts at ground level
        bs_loc = np.array([900, 100, 20])     # Base station location
        ris_loc = np.array([200, 300, 40])    # RIS location
        
        # Extract coordinates
        xb, yb, zb = bs_loc
        xr, yr, zr = ris_loc
        xt, yt = target_loc[0], target_loc[1]
        
        # Speed of light
        c = 3e8
        
        # Calculate 3D Euclidean distances
        L1 = np.sqrt((xb - xr)**2 + (yb - yr)**2 + (zb - zr)**2)
        L2 = np.sqrt((xr - xt)**2 + (yr - yt)**2 + zr**2)
        L3 = np.sqrt((xb - xt)**2 + (yb - yt)**2 + zb**2)
        
        # Calculate 2D (X-Y plane) projections
        L_proj1 = np.sqrt((xb - xr)**2 + (yb - yr)**2)
        L_proj2 = np.sqrt((xr - xt)**2 + (yr - yt)**2)
        L_proj3 = np.sqrt((xb - xt)**2 + (yb - yt)**2)
        
        # Calculate signal delays
        delays = {
            'line_of_sight': L3 / c,
            'non_line_of_sight': (L1 + L2) / c
        }
        
        # Calculate angles
        angles = {
            'bs_to_ris': {
                'azimuth': np.arcsin((zb - zr) / L1),
                'elevation_azimuth': np.arcsin((xb - xr) / L_proj1),
                'elevation_angle': np.arccos((zb - zr) / L1)
            },
            'ris_to_target': {
                'aoa': np.arcsin(zr / L2),
                'azimuth': np.arccos((yr - yt) / L_proj2),
                'elevation_angle': np.arccos(zr / L2)
            },
            'bs_to_target': {
                'aoa_in': np.arccos(zb/L3),
                'aoa_out': np.arcsin(zb/L3)
            }
        }
        
        return L1, L2, L3, L_proj1, L_proj2, L_proj3, delays, angles
    
    # TODO: Implement PEB generation function here
    # TODO: Implement state function to derive reward from PEB
    # TODO: Implement action function to act on that reward changing  
    
    
def simulate_ris_v2x():
    WirelessChannelSimulation